runtime/lua/vim/treesitter/languagetree.lua

```diff
@@ -409,6 +423,41 @@ function LanguageTree:_add_injections()
   return query_time
 end
 
+--- Run an asynchronous parse, calling {on_parse} when complete.
+---
+--- @param range boolean|Range?
+--- @param on_parse fun(trees?: table<integer, TSTree>, err?: string)
+--- @return table<integer, TSTree>? trees the list of parsed trees, if parsing completed synchronously
+function LanguageTree:_async_parse(range, on_parse)
+  local ct = self:_buf().changedtick
+  local total_parse_time = 0
+  local redrawtime = vim.o.redrawtime
+
+  local function step()
+    -- If buffer was changed in the middle of parsing, reset parse state
+    if self:_buf().changedtick ~= ct then
+      self._parser:reset()
```

Comment:
Injection parsers also need to be reset, I assume.

Tree sitted docs say reaet need to bw called every time the parse is restarted. This logic should probably happen when the tree is invalidated (because of this: https://github.com/neovim/neovim/pull/31631#issuecomment-2562135903)